compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala

```diff
@@ -245,51 +245,36 @@ object PrepareInlineable {
-          isMacro = true
-          splice
-        }
-        override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree =
-          if (isMacro) tree else super.transform(tree)
-      }.transform(rhs)
-
-      if (isMacro) {
-        sym.setFlag(Macro)
-        if (level == 0) {
-          def isValidMacro(tree: Tree)(implicit ctx: Context): Unit = tree match {
-            case Spliced(code) =>
-              if (code.symbol.flags.is(Inline))
-                ctx.error("Macro cannot be implemented with an `inline` method", code.sourcePos)
-              Splicer.checkValidMacroBody(code)
-              new PCPCheckAndHeal(freshStagingContext).transform(rhs) // Ignore output, only check PCP
+    if (sym.is(Macro) && !ctx.isAfterTyper) {
+      def isValidMacro(tree: Tree)(implicit ctx: Context): Unit = tree match {
+        case Spliced(code) =>
+          if (code.symbol.flags.is(Inline))
+            ctx.error("Macro cannot be implemented with an `inline` method", code.sourcePos)
+          Splicer.checkValidMacroBody(code)
+          new PCPCheckAndHeal(freshStagingContext).transform(rhs) // Ignore output, only check PCP
 
-            case Block(List(stat), Literal(Constants.Constant(()))) => isValidMacro(stat)
-            case Block(Nil, expr) => isValidMacro(expr)
-            case Typed(expr, _) => isValidMacro(expr)
-            case Block(DefDef(nme.ANON_FUN, _, _, _, _) :: Nil, Closure(_, fn, _)) if fn.symbol.info.isImplicitMethod =>
-              // TODO Suppot this pattern
-              ctx.error(
-                """Macros using a return type of the form `foo(): given X => Y` are not yet supported.
-                  |
-                  |Place the implicit as an argument (`foo() given X: Y`) to overcome this limitation.
-                  |""".stripMargin, tree.sourcePos)
-            case _ =>
-              ctx.error(
-                """Malformed macro.
-                  |
-                  |Expected the splice ${...} to be at the top of the RHS:
-                  |  inline def foo(inline x: X, ..., y: Y): Int = ${impl(x, ... '{y}})
-                  |
-                  | * The contents of the splice must call a static method
-                  | * All arguments must be quoted or inline
-                """.stripMargin, pos)
-          }
-          isValidMacro(rhs)
-        }
+        case Block(List(stat), Literal(Constants.Constant(()))) => isValidMacro(stat)
+        case Block(Nil, expr) => isValidMacro(expr)
+        case Typed(expr, _) => isValidMacro(expr)
+        case Block(DefDef(nme.ANON_FUN, _, _, _, _) :: Nil, Closure(_, fn, _)) if fn.symbol.info.isImplicitMethod =>
+          // TODO Suppot this pattern
+          ctx.error(
+            """Macros using a return type of the form `foo(): given X => Y` are not yet supported.
+              |
+              |Place the implicit as an argument (`foo() given X: Y`) to overcome this limitation.
```

Comment:
Maybe just `Write ... to overcome this problem` to avoid terminology change about implicits.

Reply:
I will support this pattern soon. The error message will be removed then.